test/tbb/test_join_node_preview.cpp

```diff
@@ -82,6 +82,149 @@ void test_follows_and_precedes_api() {
+
+        g.wait_for_all();
+
+        for (auto item : new_work_items) {
+            CHECK_MESSAGE(processed_items[check_index++] == item, "Unexpected start_work_items processing");
+        }
+    });
+}
+
+void test_try_put_and_wait_reserving() {
+    tbb::task_arena arena;
+
+    arena.execute([]{
+        tbb::flow::graph g;
+
+        std::vector<int> start_work_items;
+        std::vector<int> processed_items;
+        std::vector<int> new_work_items;
+        int wait_message = 10;
+
+        for (int i = 0; i < wait_message; ++i) {
+            start_work_items.emplace_back(i);
+            new_work_items.emplace_back(i + 1 + wait_message);
+        }
+
+        using tuple_type = std::tuple<int, int, int>;
+        tbb::flow::queue_node<int> buffer1(g);
+        tbb::flow::queue_node<int> buffer2(g);
+        tbb::flow::queue_node<int> buffer3(g);
+
+        tbb::flow::join_node<tuple_type, tbb::flow::reserving> join(g);
+
+        tbb::flow::function_node<tuple_type, int, tbb::flow::rejecting> function(g, tbb::flow::serial,
+            [&](tuple_type tuple) noexcept {
+                CHECK(std::get<0>(tuple) == std::get<1>(tuple));
+                CHECK(std::get<1>(tuple) == std::get<2>(tuple));
+
+                auto input = std::get<0>(tuple);
+
+                if (input == wait_message) {
+                    for (auto item : new_work_items) {
+                        buffer1.try_put(item);
+                        buffer2.try_put(item);
+                        buffer3.try_put(item);
+                    }
+                }
+                processed_items.emplace_back(input);
+                return 0;
+            });
+
+        tbb::flow::make_edge(buffer1, tbb::flow::input_port<0>(join));
+        tbb::flow::make_edge(buffer2, tbb::flow::input_port<1>(join));
+        tbb::flow::make_edge(buffer3, tbb::flow::input_port<2>(join));
+        tbb::flow::make_edge(join, function);
+
+        for (auto item : start_work_items) {
+            buffer1.try_put(item);
+            buffer2.try_put(item);
+            buffer3.try_put(item);
+        }
+
+        buffer1.try_put(wait_message);
+        buffer2.try_put(wait_message);
+        buffer3.try_put_and_wait(wait_message);
+
+        // It is expected that the join_node would push the tuple of three copies of first element in start_work_items
+        // And occupy the concurrency of function. Other tuples would be rejected and taken using push-pull protocol
+        // between function and join_node and between join_node and each buffer in FIFO order because queue_node is used
+        std::size_t check_index = 0;
+
+        for (auto item : start_work_items) {
+            CHECK_MESSAGE(processed_items[check_index++] == item, "Unexpected start_work_items processing");
+        }
+
+        CHECK_MESSAGE(processed_items[check_index++] == wait_message, "Unexpected wait_message processing");
```

Comment:
When I add this CHECK it fails for me. Is that expected?  

I see:

values: CHECK( 11 == 21 )

Reply:
There was an issue in the reserving node test - the `task_arena` was default initialized instead of explicitly specifying `1` as a concurrency limit. Because of this, the test sporadically failed. Now it is fixed.
Thanks for catching that.